### PR TITLE
Add support for YAML configuration file format

### LINKED ISSUE #1109: Add contribution model for application configuration
I would love to be able to create a new project that contains an `application.yml` instead of an `application.properties` file.

Maybe as an option? `properties=yaml`

Or maybe `whiteSpaceBugs=true` :trollface: 
Joking aside, we commonly give instructions to rename app.props to app.yml after creating a new project and then provide example configuration in YAML format.

I can understand the desire to keep the number of permutations down, so feel free to close this issue!

Thanks again!

### PR DESCRIPTION
**Pull request by YangSiJun528**
This PR adds an option to generate project configuration in YAML format. Currently, only `.properties` is supported. (#1109)

This PR has many changes, so I'm opening this PR early to get feedback before completing all implementation details. The core features work well, but I still need to add more comprehensive tests - currently only basic tests are implemented.

## Main Changes
- Add `ConfigurationFileFormat` interface with factory and implementations for Properties and YAML
- Update `ApplicationProperties` to write YAML format  
- Add `ApplicationYamlPropertiesContributor`
- Add `@ConditionalOnConfigurationFileFormat` and `OnConfigurationFileFormatCondition` to choose between Properties and YAML
- Add config file format support to `MutableProjectDescription`, `InitializrMetadata`, and `ProjectRequest`
- Update tests: since configuration file format is now configurable, remove `application.properties` validation or set format explicitly where needed.

## Need Feedback On
- Is the name and package location of `ConfigurationFileFormat` appropriate? (`io.spring.initializr.generator.configuration.format`)
- Any feedback on the overall design? 
- Are there any missing features or edge cases to consider?

Once I receive feedback on the approach, I will add comprehensive tests and polish the implementation before requesting final review.

---

Fixes gh-1109

### CODE REVIEW

**Comment by mhalbritter** [comment]
Thanks @YangSiJun528, I'm going to take a look. I'm in the middle of something right now, but i plan to look at it this week.

File: initializr-web/src/test/resources/metadata/test-default-2.0.0-ssl.json [hunk 1]
```diff
@@ -149,6 +149,20 @@
       }
     ]
   },
+  "configurationFileFormat": {
```
Comment (mhalbritter) [inline initializr-web/src/test/resources/metadata/test-default-2.0.0-ssl.json:152]: This changes the metadata v2.0.0 format. I think we should add a new metadata version v2.3 which adds the `configurationFileFormat` field. You can take a look at `InitializrMetadataV22JsonMapper` and `InitializrMetadataV21JsonMapper` for that.

**Comment by mhalbritter** [comment]
Hi @YangSiJun528 ,

thanks for the PR. I've left some comments for your consideration.

I think your name `ConfigurationFileFormat` is a good one. Aside from the new metadata version the design looks sane, and I can't see any missing edge cases.

File: initializr-web/src/test/java/io/spring/initializr/web/controller/ProjectGenerationControllerIntegrationTests.java [hunk 2]
```diff
@@ -53,7 +53,37 @@ void simpleZipProject() {
...
+		assertThat(project).mavenBuild()
+			.hasDependenciesSize(2)
+			.hasDependency(Dependency.createSpringBootStarter("web"))
+			.hasDependency(Dependency.createSpringBootStarter("test", Dependency.SCOPE_TEST));
+		assertThat(project).containsFiles("src/main/resources/application.properties");
+	}
 
+	@Test
+	void simpleZipProjectWithYAMLFormat() {
+		ResponseEntity<byte[]> entity = downloadArchive("/starter.zip?dependencies=web&&configurationFileFormat=yaml");
```
[the remark this answers is not in the document]
Comment (YangSiJun528) [reply initializr-web/src/test/java/io/spring/initializr/web/controller/ProjectGenerationControllerIntegrationTests.java]: I fixed it.

File: initializr-web/src/test/resources/metadata/test-default-2.0.0-ssl.json (hunk 1 above)
Comment (YangSiJun528) [reply initializr-web/src/test/resources/metadata/test-default-2.0.0-ssl.json]: I reverted the changes to `InitializrMetadataV2JsonMapper` and added `InitializrMetadataV23JsonMapper`.

**Comment by YangSiJun528** [comment]
Thanks for making this change!

I initially used `.yml` in this PR because Spring Initializr does, and in my experience, `.yml` is more common in Spring projects. Since the Spring Boot docs mix `.yaml` and `.yml`, I assumed `.yml` would be preferred.

After looking into it, I learned [`.yaml` is officially recommended by the YAML spec](https://yaml.org/faq.html), so I agree using `.yaml` makes sense.

Just leaving a note in case anyone wonders why `.yml` wasn’t used.
